### DISCUSSION TITLE
How to load homebrew on Fish shell?

### DISCUSSION THREAD

**Discussion by andradei** [Homebrew/discussions#4412] [Getting started] [answered]
### Output of `brew config`

```shell
Doesn't show until it `brew` is loaded.
```

### Output of `brew doctor`

```shell
Doesn't show until it `brew` is loaded.
```

### Description of issue

It would be nice to have documentation showing how to run homebrew on the Fish shell.
I can run the installation script, but don't know how to run `brew shellenv` in `config.fish` to load it.

**Comment by SMillerDev** [comment]
It should suggest it already in the install command. But if it doesn't, it's exactly the same as for bash, except you need to do `eval (` instead of `eval $(`

**Reply by BarbzYHOOL** [reply]
no, fish doesn't read .profile file

**Comment by apainintheneck** [comment]
Which OS are you trying to install Homebrew on?

On Intel Macs, the brew executable got symlinked into `/usr/local/bin` so no changes were necessary since that's already in everyone's path.

On newer M1 Macs, I ended up adding the following to my `config.fish` file. It should be suggested in the install script.

```fish
eval "$(/opt/homebrew/bin/brew shellenv)"
```

The install script suggests using `.profile` by default if the user is not using `bash` or `zsh`.

https://github.com/Homebrew/install/blob/35c494693c6d5640bc9892bfcac7eb2433051c7c/install.sh#L1014-L1051

It'd probably make sense to add information about `fish` too so that the user knows to add it to `~/.config/fish/config.fish`. This seems to be the expected place to put those sorts of shell startup scripts and is where `asdf` and `ghcup` put that information for `fish` users for instance.

- https://fishshell.com/docs/current/index.html#configuration

**Reply by apainintheneck** [reply]
Opened a PR to change it: https://github.com/Homebrew/install/pull/811

**Reply by BarbzYHOOL** [reply]
for me it was on linux

**Reply by apainintheneck** [reply]
The default prefix on Linux is `/home/linuxbrew/.linuxbrew` so the following probably works.

```sh
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
```

**Reply by BarbzYHOOL** [reply]
what about putting it in /etc/environment ??

**Reply by paldepind** [reply]
Note that this only works in Fish 3.4 and greater. For versions below the following can be used (which is the more idiomatic way of doing it anyway):

```fish
eval (/home/linuxbrew/.linuxbrew/bin/brew shellenv)
```

**Reply by aldoKelvianto** [reply]
Agree with @paldepind, the command substitution in Fish is using `()`, not `"$()"`.

https://fishshell.com/docs/current/tutorial.html#command-substitutions

**Comment by mattmc3** [comment]
The proper shell command for Fish is simply `brew shellenv fish | source`. If `fish_add_path` has not been called so that `$PATH` contains the location of of brew, you can simply use something like this:

```fish
set brewcmd (path filter /opt/homebrew/bin/brew /usr/local/bin/brew)[1]
and $brewcmd shellenv fish | source
```

**Comment by andradei** [comment, marked as answer]
This is what I have been using without problems for months after asking this question:

```fish
if test -d /home/linuxbrew/.linuxbrew
      # Homebrew is installed on Linux

      set -gx HOMEBREW_PREFIX "/home/linuxbrew/.linuxbrew"
      set -gx HOMEBREW_CELLAR "/home/linuxbrew/.linuxbrew/Cellar"
      set -gx HOMEBREW_REPOSITORY "/home/linuxbrew/.linuxbrew/Homebrew"
      set -gx PATH "/home/linuxbrew/.linuxbrew/bin" "/home/linuxbrew/.linuxbrew/sbin" $PATH
      set -q MANPATH; or set MANPATH ''
      set -gx MANPATH "/home/linuxbrew/.linuxbrew/share/man" $MANPATH
      set -q INFOPATH; or set INFOPATH ''
      set -gx INFOPATH "/home/linuxbrew/.linuxbrew/share/info" $INFOPATH

      # Homebrew asked for this in order to `brew upgrade`
      set -gx HOMEBREW_GITHUB_API_TOKEN {api token goes here, don't remember where that's created}
  else if test -d /opt/homebrew
      # Homebrew is installed on MacOS

      /opt/homebrew/bin/brew shellenv | source
  end
end
```

**Reply by kashrobinson71-tech** [reply]
Thanks

**Reply by gromgit** [reply]
Note that the above advice is outdated. `brew shellenv | source` works on all supported platforms, no need for special Linux handling.

**Comment by tmwatchanan** [comment]
This might be a little bit better.

```fish
if test -d /home/linuxbrew/.linuxbrew # Linux
	set -gx HOMEBREW_PREFIX "/home/linuxbrew/.linuxbrew"
	set -gx HOMEBREW_CELLAR "$HOMEBREW_PREFIX/Cellar"
	set -gx HOMEBREW_REPOSITORY "$HOMEBREW_PREFIX/Homebrew"
else if test -d /opt/homebrew # MacOS
	set -gx HOMEBREW_PREFIX "/opt/homebrew"
	set -gx HOMEBREW_CELLAR "$HOMEBREW_PREFIX/Cellar"
	set -gx HOMEBREW_REPOSITORY "$HOMEBREW_PREFIX/homebrew"
end
fish_add_path -gP "$HOMEBREW_PREFIX/bin" "$HOMEBREW_PREFIX/sbin";
! set -q MANPATH; and set MANPATH ''; set -gx MANPATH "$HOMEBREW_PREFIX/share/man" $MANPATH;
! set -q INFOPATH; and set INFOPATH ''; set -gx INFOPATH "$HOMEBREW_PREFIX/share/info" $INFOPATH;
```

**Reply by kashrobinson71-tech** [reply]
Thanks

**Comment by kode54** [comment]
Thanks for the fish method, since I've "brilliantly" decided to use the Homebrew version of Fish as my default shell on macOS.

**Comment by jiawen** [comment]
On macOS and Homebrew 4.6.11, `brew shellenv` seems to recognize `fish` as my shell and export the right thing:

```
❯ /opt/homebrew/bin/brew shellenv
set --global --export HOMEBREW_PREFIX "/opt/homebrew";
set --global --export HOMEBREW_CELLAR "/opt/homebrew/Cellar";
set --global --export HOMEBREW_REPOSITORY "/opt/homebrew";
fish_add_path --global --move --path "/opt/homebrew/bin" "/opt/homebrew/sbin";
if test -n "$MANPATH[1]"; set --global --export MANPATH '' $MANPATH; end;
if not contains "/opt/homebrew/share/info" $INFOPATH; set --global --export INFOPATH "/opt/homebrew/share/info" $INFOPATH; end;
```

**Reply by jiawen** [reply]
And on Linux as well:

```
❯ /home/linuxbrew/.linuxbrew/bin/brew shellenv
set --global --export HOMEBREW_PREFIX "/home/linuxbrew/.linuxbrew";
set --global --export HOMEBREW_CELLAR "/home/linuxbrew/.linuxbrew/Cellar";
set --global --export HOMEBREW_REPOSITORY "/home/linuxbrew/.linuxbrew/Homebrew";
fish_add_path --global --move --path "/home/linuxbrew/.linuxbrew/bin" "/home/linuxbrew/.linuxbrew/sbin";
if test -n "$MANPATH[1]"; set --global --export MANPATH '' $MANPATH; end;
if not contains "/home/linuxbrew/.linuxbrew/share/info" $INFOPATH; set --global --export INFOPATH "/home/linuxbrew/.linuxbrew/share/info" $INFOPATH; end;
```

**Reply by kode54** [reply]
Yes, that emits the right sequence of commands. I was missing the part of how to source them into my fish session.
